### ISSUE TITLE
chezmoi thinks that all files have been modified

### ISSUE THREAD

**Issue by twpayne** [twpayne/chezmoi#1066]
## Describe the bug

This was reported by @fwSmit in #1039, specifically in [this comment](https://github.com/twpayne/chezmoi/issues/1039#issuecomment-798515391):

> I ran into this bug when running `chezmoi2 apply --source-path ...`. It gave an error that no tty was found, suggesting that some user input was needed. It turns out, some entirely different file was changed according to chezmoi2, even though I didn't change it.
> 
> ```shell
> ❯ chezmoi2 apply
> .XCompose has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? diff
> diff --git a/.XCompose b/.XCompose
> .XCompose has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? quit
> ❯ diff ./.XCompose ~/.local/share/chezmoi/dot_XCompose
> ❯ ls -l .XCompose
> -rw-r--r-- 1 friso friso 3413 Sep 19 08:57 .XCompose
> ❯ ls -l ~/.local/share/chezmoi/dot_XCompose
> -rw-r--r-- 1 friso friso 3413 Sep 19 08:57 ~/.local/share/chezmoi/dot_XCompose
> ```
> 
> I don't see any changes between the source file and the destination file, so maybe the internal cache is corrupted or something.
> 
> ```
> .XCompose has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? skip
> .config/autostart has changed since chezmoi last wrote it [overwrite,all-overwite,skip,quit]? skip
> ...
> .config/git/config has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? skip
> .config/gtk-3.0/colors.css has changed since chezmoi last wrote it [diff,overwrite,all-overwite,skip,quit]? quit
> ```
> 
> It seems like all files "have been changed since chezmoi last wrote it".
> I consider this to be two bugs:
> 
> * Chezmoi incorrectly says the file has changed.
> * `--source-path` should ignore all other files

Following conversation:

> Thanks for reporting this @fwSmit.
> 
> > * `--source-path` should ignore all other files
> 
> I think this is already the case, I tried to reproduce the problem with [ebdd6f4](https://github.com/twpayne/chezmoi/commit/ebdd6f4eb239c5b99fc22a8f9720ae8635ed9a62), but was unable to do so.
> 
> > Chezmoi incorrectly says the file has changed.
> 
> Given that chezmoi reports that all files have been modified, when I suspect that what's happened here is that chezmoi's state has been corrupted somehow. Would you mind running the following commands?
> 
> ```
> sha256sum ~/.XCompose
> sha256sum ~/.local/share/chezmoi/dot_XCompose
> chezmoi state dump | grep -A4 XCompose
> stat ~/.XCompose
> ```
> 
> This should help reveal where the problem is.

> > I think this is already the case, I tried to reproduce the problem with [ebdd6f4](https://github.com/twpayne/chezmoi/commit/ebdd6f4eb239c5b99fc22a8f9720ae8635ed9a62), but was unable to do so.
> 
> My bad, I was still on an older version and didn't test all bugs on the up-to-date version. The other bug still persists with the new, version however.
> 
> ```shell
> > sha256sum ~/.XCompose
> sha256sum ~/.local/share/chezmoi/dot_XCompose
> ff0bca2e72cb19207e94969741ddf49fe31ed7b72ed5fa6813e663466364bb80  /home/friso/.XCompose
> ff0bca2e72cb19207e94969741ddf49fe31ed7b72ed5fa6813e663466364bb80  /home/friso/.local/share/chezmoi/dot_XCompose
> 
> > chezmoi2 state dump | grep -A4 XCompose 
>    "/home/friso/.XCompose": {
>       "contentsSHA256": "ff0bca2e72cb19207e94969741ddf49fe31ed7b72ed5fa6813e663466364bb80",
>       "mode": 438,
>       "type": "file"
>     },
> 
> > stat ~/.XCompose
>   File: /home/friso/.XCompose
>   Size: 3413      	Blocks: 8          IO Block: 4096   regular file
> Device: 10305h/66309d	Inode: 5768362     Links: 1
> Access: (0644/-rw-r--r--)  Uid: ( 1000/   friso)   Gid: ( 1000/   friso)
> Access: 2021-03-13 19:27:07.090972357 +0100
> Modify: 2020-09-19 08:57:05.303266433 +0200
> Change: 2020-09-19 08:57:05.309933080 +0200
>  Birth: 2020-09-19 08:57:05.303266433 +0200
> ```
> 
> I don't see anything weird here, but maybe you do

> > I don't see anything weird here, but maybe you do
> 
> @fwSmit it looks like the sha256 sums all line up. Is it possible that there's been a `umask` change? What does the output of `umask` print?

> Umask just prints 022

## To reproduce

We don't have a reproduction case yet. @fwSmit are you able to shed any more light on how to reproduce this?

## Expected behavior

chezmoi should not give false positives when files have not changed.

**Comment by fwSmit** [comment]
What makes it a bit more difficult is that I was using an older version of chezmoi2, so I'm not entirely sure if it's completely reproducible in the new version. (to be clear the bug is still present, but we don't know if you can get the db somehow corrupted in the new version).
The only thing I can remember doing with chezmoi2 to create the bug is letting vim run the it on a single file with the  source-path argument. Is there anything in the db that could help. I'm willing to share it privately

**Comment by fwSmit** [comment]
I was a bit hesitant to do it for investigation sake and because I wasn't sure what was gonna be written, but choosing `overwrite-all` at least solves the issue. I don't know how it is caused though. I have made a backup of the chezmoi db, so again, if you need it just tell me.

**Comment by sheeley** [comment]
I've been encountering this with my `~/.config` directory - I'm unsure if it's related to this or should be a separate issue, please let me know if you'd like me to split this off.
```
.config has changed since chezmoi last wrote it [overwrite,all-overwrite,skip,quit]? overwrite
```
I'm not sure what could be different, when I look at the filesystem, I see this both before and after `chezmoi apply` (`--verbose` doesn't add any detail here:
```
drwxr-xr-x  11 sheeley  staff   352B Apr  6 06:48 .config
```
This is the version I'm running:
```
chezmoi version v2.0.9, commit 7c7213b32435823ce3fd0dc468fdecefe7c5e8f9, built at 2021-04-04T20:20:16Z, built by homebrew
```
And `chezmoi doctor` looks good:
```
RESULT   CHECK                MESSAGE
ok       version              v2.0.9, commit 7c7213b32435823ce3fd0dc468fdecefe7c5e8f9, built at 2021-04-04T20:20:16Z, built by homebrew
ok       os-arch              darwin/amd64
ok       config-file          /Users/sheeley/.config/chezmoi/chezmoi.toml is a file
ok       source-dir           /Users/sheeley/.local/share/chezmoi is a directory
ok       suspicious-entries   no suspicious entries found in /Users/sheeley/.local/share/chezmoi
ok       dest-dir             /Users/sheeley is a directory
ok       shell                found /usr/local/bin/fish
ok       editor               found /usr/local/bin/code
ok       git-cli              found /usr/local/bin/git, version 2.31.1
ok       merge-cli            found /usr/bin/vimdiff
ok       age-cli              age not found in $PATH
ok       gnupg-cli            gpg not found in $PATH
info     1password-cli        op not found in $PATH
info     bitwarden-cli        bw not found in $PATH
info     gopass-cli           gopass not found in $PATH
info     keepassxc-cli        keepassxc-cli not found in $PATH
info     keepassxc-db         not set
info     lastpass-cli         lpass not found in $PATH
info     pass-cli             pass not found in $PATH
info     vault-cli            vault not found in $PATH
info     secret-cli           not set
```

**Comment by twpayne** [comment]
Should be fixed with #1149.

**Comment by sheeley** [comment]
Just installed dev and I believe I can confirm that for my use case 👍 Thanks, @twpayne!
